mempool/src/shared_mempool/priority.rs

```diff
@@ -208,11 +242,149 @@ impl PrioritizedPeersState {
+            let peers_in_vfn_network = self
+                .prioritized_peers
+                .read()
+                .iter()
+                .cloned()
+                .filter(|peer| peer.network_id() == NetworkId::Vfn)
+                .collect::<Vec<_>>();
+
+            if !peers_in_vfn_network.is_empty() {
+                top_peers = vec![peers_in_vfn_network[0]];
+            }
+        }
+
+        if top_peers.is_empty() {
+            let base_ping_latency = self.prioritized_peers.read().first().and_then(|peer| {
+                peer_monitoring_data
+                    .get(peer)
+                    .and_then(|metadata| get_peer_ping_latency(metadata))
+            });
+
+            // Extract top peers with ping latency less than base_ping_latency + 30 ms
+            for peer in self.prioritized_peers.read().iter() {
+                if top_peers.len() >= num_top_peers as usize {
+                    break;
+                }
+
+                let ping_latency = peer_monitoring_data
+                    .get(peer)
+                    .and_then(|metadata| get_peer_ping_latency(metadata));
+
+                if base_ping_latency.is_none()
+                    || ping_latency.is_none()
+                    || ping_latency.unwrap()
+                        < base_ping_latency.unwrap()
+                            + (self.mempool_config.latency_slack_between_top_upstream_peers as f64)
+                                / 1000.0
+                {
+                    top_peers.push(*peer);
+                }
+            }
+        }
+        info!(
+            "Identified top peers: {:?}, node_type: {:?}",
+            top_peers, self.node_type
+        );
+
+        assert!(top_peers.len() <= num_top_peers as usize);
+        // Top peers shouldn't be empty if prioritized_peers is not zero
+        assert!(self.prioritized_peers.read().is_empty() || !top_peers.is_empty());
+
+        self.peer_to_sender_buckets = HashMap::new();
+        if !self.prioritized_peers.read().is_empty() {
+            // Assign sender buckets with Primary priority
+            let mut peer_index = 0;
+            for bucket_index in 0..self.mempool_config.num_sender_buckets {
+                self.peer_to_sender_buckets
+                    .entry(*top_peers.get(peer_index).unwrap())
+                    .or_default()
+                    .insert(bucket_index, BroadcastPeerPriority::Primary);
+                peer_index = (peer_index + 1) % top_peers.len();
+            }
+
+            // Assign sender buckets with Failover priority. Use Round Robin.
+            peer_index = 0;
+            for _ in 0..self.mempool_config.default_failovers {
+                for bucket_index in 0..self.mempool_config.num_sender_buckets {
+                    // Find the first peer that already doesn't have the sender bucket, and add the bucket
+                    for _ in 0..self.prioritized_peers.read().len() {
```

Comment:
should you do read() only once? 

should we randomize self.prioritized_peers.read() order, or randomly pick the index for each?

do we need to make sure that fallback is different from primary, for each of the sender buckets?

Reply:
> should we randomize self.prioritized_peers.read() order

Some of the upstream peers have super high ping latency (like 250 ms). So, I'm still prioritizing based on latency.

> do we need to make sure that fallback is different from primary, for each of the sender buckets?

Yes. We need the Failover to be different from Primary. We have Failovers so that the system still runs when the Primary fails.